File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
private:
    unordered_map<int, int> valToIndex;
    vector<int> values;
public:
    RandomizedSet() {
        srand(time(nullptr));
    }
    
    bool insert(int val) {
        if (valToIndex.find(val) != valToIndex.end()) {
            return false;
        }
        values.push_back(val);
        valToIndex[val] = values.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        auto it = valToIndex.find(val);
        if (it == valToIndex.end()){
            return false;
        }
        int lastElement = values.back();
        int idx = it->second;
        values[idx] = lastElement;
        valToIndex[lastElement] = idx;
        values.pop_back();
        valToIndex.erase(it);
        return true;
    }
    
    int getRandom() {
        int randomIdx = rand() % values.size();
        return values[randomIdx];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.rs
use std::collections::HashMap;

struct RandomizedSet {
    val_to_index: HashMap<i32, usize>,
    values: Vec<i32>, 
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RandomizedSet {

    fn new() -> Self {
        RandomizedSet {
            val_to_index: HashMap::new(),
            values: Vec::new(),
        }
    }
    
    fn insert(&mut self, val: i32) -> bool {
        if self.val_to_index.contains_key(&val) {
            return false;
        }
        self.values.push(val);
        self.val_to_index.insert(val, self.values.len() - 1);
        true
    }
    
    fn remove(&mut self, val: i32) -> bool {
        if let Some(&index) = self.val_to_index.get(&val) {
            let last_element = *self.values.last().unwrap();
            self.values[index] = last_element;
            self.val_to_index.insert(last_element, index);
            self.values.pop();
            self.val_to_index.remove(&val);
            true
        } else {
            false
        }
    }
    
    fn get_random(&self) -> i32 {
        use rand::Rng;
        let mut rng = rand::thread_rng();
        let random_index = rng.gen_range(0..self.values.len());
        self.values[random_index]
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * let obj = RandomizedSet::new();
 * let ret_1: bool = obj.insert(val);
 * let ret_2: bool = obj.remove(val);
 * let ret_3: i32 = obj.get_random();
 */